File: server/apps/pages/api/serializers/personal.py
from rest_framework import serializers
from ...models.personal import Personal
from .personal_social_link import PersonalSocialLinkSerializer
from .skill import PersonalSkillSerializer
from ..mixins import TranslationMixin


class PersonalSerializer(TranslationMixin, serializers.ModelSerializer):
    social = PersonalSocialLinkSerializer(source='social_links', many=True, read_only=True)
    skills = PersonalSkillSerializer(many=True, read_only=True)
    name = serializers.SerializerMethodField()
    role = serializers.SerializerMethodField()
    bio = serializers.SerializerMethodField()

    def get_name(self, obj) -> str:
        return self.get_translated_field(obj, "name")

    def get_role(self, obj) -> str:
        return self.get_translated_field(obj, "role")
    
    def get_bio(self, obj) -> str:
        return self.get_translated_field(obj, "bio")

    class Meta:
        model = Personal
        fields = (
            "id",
            "name",
            "role",
            "bio",
            "poster",
            "email",
            "phone",
            "age",
            "blood_group",
            "website",
            "address",
            "social",
            "skills",
            "slug",
            "created_at",
        )
        read_only_fields = (
            "id",
            "created_at",
            "social",
            "skills",
            "slug",
        )

File: intro-app/src/api/modules/personal.js
import { publicApiClient } from "../client";
import { API_ENDPOINTS, API_BASE_URL } from "../../constants/constants.js";

const normalizePersonalData = (item) => {
    if (!item) return null;

    const baseUrl = API_BASE_URL.replace('/api/', '');

    const getFullUrl = (url) => {
        if (!url) return null;
        if (url.startsWith('http')) return url;
        return url.startsWith('/') ? `${baseUrl}${url}` : `${baseUrl}/${url}`;
    };

    // Normalize skills - check if skills exist, otherwise create from skill fields
    let skills = item.skills || [];

    // If no skills array but individual skill fields exist, create skills array
    if (skills.length === 0) {
        const skillFields = [];

        // Check for skill_1, skill_2, skill_3 pattern
        if (item.skill_1_name && item.skill_1_value) {
            skillFields.push({ name: item.skill_1_name, value: item.skill_1_value });
        }
        if (item.skill_2_name && item.skill_2_value) {
            skillFields.push({ name: item.skill_2_name, value: item.skill_2_value });
        }
        if (item.skill_3_name && item.skill_3_value) {
            skillFields.push({ name: item.skill_3_name, value: item.skill_3_value });
        }

        if (skillFields.length > 0) {
            skills = skillFields;
        }
    }

    return {
        ...item,
        id: item.id,
        name: item.name || '',
        role: item.role || item.position || '',
        position: item.role || item.position || '',
        bio: item.bio || '',
        description: item.bio || '',
        email: item.email || '',
        phone: item.phone || '',
        age: item.age || null,
        blood_group: item.blood_group || '',
        website: item.website || '',
        address: item.address || '',
        poster: getFullUrl(item.poster || item.image || item.avatar || item.photo),
        image: getFullUrl(item.poster || item.image || item.avatar || item.photo),
        avatar: getFullUrl(item.poster || item.image || item.avatar || item.photo),
        slug: item.slug || item.id,
        social: item.social || [],
        skills: skills
    };
};

export const PersonalAPI = {
    async getAll() {
        try {
            const { data } = await publicApiClient.get(
                API_ENDPOINTS.modules.personal.index
            );

            const items = data?.data || data || [];

            if (!Array.isArray(items)) {
                return [];
            }

            return items.map(normalizePersonalData).filter(Boolean);
        } catch (error) {
            throw error;
        }
    },

    async getBySlug(slug) {
        try {
            const { data } = await publicApiClient.get(
                `${API_ENDPOINTS.modules.personal.index}/${slug}/`
            );

            const memberData = data?.data || data;
            return normalizePersonalData(memberData);
        } catch (error) {
            throw error;
        }
    }
};
